### PR TITLE
Add proc-macro related config and tests

### PR DESCRIPTION
**Pull request by edwin0cheng**
This PR do the following things:

1. Add cli argument `proc-macro` for running proc-macro server.
2. Added support for proc-macro in bench and analysis-stats
3. Added typescript config for proc-macros
4. Added an heavy test for proc-macros. 

To test it out: 

1. run `cargo xtask install --proc-macro`
2. add `"rust-analyzer.cargo.loadOutDirsFromCheck": true"` and `"rust-analyzer.procMacro.enabled": true"` in vs code config.

[Edit] Change to use `rust-analyzer proc-macro` for running proc-macro standalone process.

### CODE REVIEW

File: crates/rust-analyzer/tests/heavy_tests/main.rs [hunk 1]
```diff
@@ -625,3 +625,90 @@ fn main() { message(); }
...
+    fn_stream.extend(vec![t!("fn"), t!("bar"), t!(()), t!({})]);
+    tokens.push(Group::new(Delimiter::Brace, fn_stream).into());
+    res.extend(tokens);
+    res
+}
+
+"###,
+    )
+    .with_config(|config| {
+        let macro_srv_path = std::path::Path::new(std::env!("CARGO_MANIFEST_DIR"))
```
Comment (edwin0cheng) [inline crates/rust-analyzer/tests/heavy_tests/main.rs:695]: @matklad  This is quite tricky to make test works with another binary. Do you have any better idea to do so ?

**Comment by lnicola** [comment]
Maybe it doesn't make sense, but could we put the functionality in the same binary to simplify deployment (`e.g. rust-analyzer proc-macro-server`)?

File: crates/rust-analyzer/tests/heavy_tests/main.rs [hunk 2]
```diff
@@ -625,3 +625,90 @@ fn main() { message(); }
...
+    ($n:literal) => {
+        TokenTree::from(Ident::new($n, Span::call_site()))
+    };
+    ({}) => {
+        TokenTree::from(Group::new(Delimiter::Brace, TokenStream::new()))
+    };
+    (()) => {
+        TokenTree::from(Group::new(Delimiter::Parenthesis, TokenStream::new()))
+    };
+}
```
Comment (edwin0cheng) [inline crates/rust-analyzer/tests/heavy_tests/main.rs:677]: Writing proc-macro without `syn` and `proc-macro` crates are fun !!

**Comment by edwin0cheng** [comment]
> Maybe it doesn't make sense, but could we put the functionality in the same binary to simplify deployment (e.g. rust-analyzer proc-macro-server)?

Oh... sounds possible ? I like this idea and could simplify some code too. but I have to check whether it is possible in Windows.

File: crates/rust-analyzer/tests/heavy_tests/main.rs (hunk 1 above)
Comment (matklad) [reply crates/rust-analyzer/tests/heavy_tests/main.rs:695]: The next stable will support this use-case natively: 

https://github.com/rust-lang/cargo/pull/7697

**Comment by matklad** [comment]
>Maybe it doesn't make sense, but could we put the functionality in the same binary to simplify deployment (e.g. rust-analyzer proc-macro-server)?

Yeah, I'd probably go with this way as well. Just let rust-analyzer spawn another rust-analyzer, and be careful not to create a fork-bomb accidentally :)

**Comment by edwin0cheng** [comment]
Changed to use rust-analyzer proc-macro for running proc-macro standalone process.

File: crates/rust-analyzer/tests/heavy_tests/main.rs (hunk 1 above)
Comment (edwin0cheng) [reply crates/rust-analyzer/tests/heavy_tests/main.rs:697]: Added a FIXME for that. thanks.

File: .github/workflows/ci.yaml [hunk 3]
```diff
@@ -85,6 +85,10 @@ jobs:
     - name: Compile
       run: cargo test --no-run
 
+    # We have to build rust-analyzer first for running related heavy tests
+    - name: Build rust-analyzer
+      run: cargo build -p rust-analyzer
```
Comment (matklad) [inline .github/workflows/ci.yaml:90]: I think this should not be neccessary -- cagro guarantees that the binaries are build before integration tests .

File: crates/ra_proc_macro/src/process.rs [hunk 4]
```diff
@@ -44,8 +44,9 @@ impl Drop for Process {
 }
 
 impl Process {
-    fn run(process_path: &Path) -> Result<Process, io::Error> {
+    fn run<T: AsRef<str>>(process_path: &Path, args: &[T]) -> Result<Process, io::Error> {
```
Comment (matklad) [inline crates/ra_proc_macro/src/process.rs:47]: As this wraps the underling `Command`, an appropriate rule-of-thumb would be about not needlessly restricting the interface. It's better to just pass through the `args` exactly as the `.args` function wants:

```
pub fn args<I, S>(&mut self, args: I) -> &mut Command where
    I: IntoIterator<Item = S>,
    S: AsRef<OsStr>, 
```

File: .github/workflows/ci.yaml (hunk 3 above)
Comment (edwin0cheng) [reply .github/workflows/ci.yaml]: Okay, removed,

**Pull request merged**